src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/TranscodingReadStream.cs

```diff
@@ -0,0 +1,273 @@
+                    Debug.Fail("We should never see this");
+                    throw new InvalidOperationException();
+            }
+        }
+
+        private async Task ReadInputChars(CancellationToken cancellationToken)
+        {
+            Debug.Assert(_byteBuffer.Array != null);
+            // If we had left-over bytes from a previous read, move it to the start of the buffer and read content in to
+            // the segment that follows.
+            Buffer.BlockCopy(
+                _byteBuffer.Array,
+                _byteBuffer.Offset,
+                _byteBuffer.Array,
+                0,
+                _byteBuffer.Count);
+            int readBytes =
+#if NETSTANDARD2_0
+                await _stream.ReadAsync(_byteBuffer.Array, _byteBuffer.Count, _byteBuffer.Array.Length, cancellationToken).ConfigureAwait(false);
+#else
+                await _stream.ReadAsync(_byteBuffer.Array.AsMemory(_byteBuffer.Count), cancellationToken).ConfigureAwait(false);
+#endif
+            _byteBuffer = new ArraySegment<byte>(_byteBuffer.Array, 0, _byteBuffer.Count + readBytes);
+
+            Debug.Assert(_charBuffer.Count == 0, "We should only expect to read more input chars once all buffered content is read");
+
+#if NETSTANDARD2_0
+            _decoder.Convert(_byteBuffer.Array, _byteBuffer.Offset, _byteBuffer.Count, _charBuffer.Array, 0, _charBuffer.Array.Length,
+                flush: readBytes == 0, out int bytesUsed, out int charsUsed, out _);
+#else
+            _decoder.Convert(
+                _byteBuffer.AsSpan(),
+                _charBuffer.Array,
+                flush: readBytes == 0,
+                out int bytesUsed,
+                out int charsUsed,
+                out _);
+#endif
+
+            Debug.Assert(_charBuffer.Array != null);
+
+            _byteBuffer = _byteBuffer.Slice(bytesUsed);
+            _charBuffer = new ArraySegment<char>(_charBuffer.Array, 0, charsUsed);
+        }
+
+        private static void ThrowArgumentOutOfRangeException(byte[] buffer, int offset, int count)
+        {
+            if (count < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count));
+            }
+
+            if (offset < 0 || offset >= buffer.Length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(offset));
+            }
+
+            if (buffer.Length - offset < count)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count));
+            }
+        }
+
+        public override long Seek(long offset, SeekOrigin origin)
+        {
+            throw new NotSupportedException();
+        }
+
+        public override void SetLength(long value)
+        {
+            throw new NotSupportedException();
+        }
+
+        public override void Write(byte[] buffer, int offset, int count)
+        {
+            throw new NotSupportedException();
+        }
+
+        protected override void Dispose(bool disposing)
```

Comment:
This needs to Dispose the stream passed into the ctor.  Otherwise, it's going to potentially leak HTTP connections until the finalizer gets around to cleaning things up.

But per my half-complete comments in the release/3.1 PR, you need to be careful: when you make the change to this read stream, if you also make the change to the write stream, then when you dispose of the stream in SerializeToStreamAsync in order to return the rented buffers, you'd also end up disposing the underlying stream there, which you shouldn't do.

This is generally handled by having a leaveOpen bool argument to the decorator stream's constructor, where it indicates whether Dispose on the decorator should dispose or not dispose the underlying stream.

Reply:
@pranavkm isn't this also missing from aspnet? or is the input stream being disposed somewhere else?